src/DependencyInjection/Compiler/FragmentRegistryPass.php

```diff
@@ -0,0 +1,148 @@
+use Symfony\Component\DependencyInjection\Compiler\PriorityTaggedServiceTrait;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\DependencyInjection\Definition;
+use Symfony\Component\DependencyInjection\Exception\RuntimeException;
+
+/**
+ * Fragment registry compiler pass.
+ *
+ * @author Yanick Witschi <https://github.com/toflar>
+ */
+class FragmentRegistryPass implements CompilerPassInterface
+{
+    use PriorityTaggedServiceTrait;
+
+    const FRAGMENT_REGISTRY = 'contao.fragment.registry';
+
+    const TAG_FRAGMENT_FRONTEND_MODULE = 'contao.fragment.frontend_module';
+    const TAG_FRAGMENT_PAGE_TYPE = 'contao.fragment.page_type';
+    const TAG_FRAGMENT_CONTENT_ELEMENT = 'contao.fragment.content_element';
+
+    const RENDERER_FRONTEND_MODULE = 'contao.fragment.renderer.frontend_module.delegating';
+    const RENDERER_PAGE_TYPE = 'contao.fragment.renderer.page_type.delegating';
+    const RENDERER_CONTENT_ELEMENT = 'contao.fragment.renderer.content_element.delegating';
+
+    const TAG_RENDERER_FRONTEND_MODULE = 'contao.fragment.renderer.frontend_module';
+    const TAG_RENDERER_PAGE_TYPE = 'contao.fragment.renderer.page_type';
+    const TAG_RENDERER_CONTENT_ELEMENT = 'contao.fragment.renderer.content_element';
+
+    /**
+     * @var Definition
+     */
+    private $fragmentRegistry;
+
+    /**
+     * Collect all the fragments and fragment renderers.
+     *
+     * @param ContainerBuilder $container
+     */
+    public function process(ContainerBuilder $container)
+    {
+        if (!$container->has(self::FRAGMENT_REGISTRY)) {
+            return;
+        }
+
+        $this->fragmentRegistry = $container->findDefinition(self::FRAGMENT_REGISTRY);
+
+        if (!$this->classImplementsInterface(
+            $this->fragmentRegistry->getClass(), FragmentRegistryInterface::class)
+        ) {
+            return;
+        }
+
+        foreach ([
+            self::TAG_FRAGMENT_FRONTEND_MODULE,
+            self::TAG_FRAGMENT_PAGE_TYPE,
+            self::TAG_FRAGMENT_CONTENT_ELEMENT,
+        ] as $tag) {
+            $this->registerFragments($container, $tag);
+        }
+
+        foreach ([
+            self::RENDERER_FRONTEND_MODULE => self::TAG_RENDERER_FRONTEND_MODULE,
+            self::RENDERER_PAGE_TYPE => self::TAG_RENDERER_PAGE_TYPE,
+            self::RENDERER_CONTENT_ELEMENT => self::TAG_FRAGMENT_CONTENT_ELEMENT,
+        ] as $renderer => $tag) {
+            $this->registerFragmentRenderers($container, $renderer, $tag);
+        }
+    }
+
+    /**
+     * @param ContainerBuilder $container
+     * @param string           $tag
+     */
+    private function registerFragments(ContainerBuilder $container, string $tag)
+    {
+        $fragments = $this->findAndSortTaggedServices($tag, $container);
+
+        foreach ($fragments as $priority => $reference) {
+            $fragment = $container->findDefinition($reference);
+            $fragmentOptions = $fragment->getTag($tag)[0];
```

Comment:
what if one fragment has multiple tags? E.g. one controller with multiple actions for different types?

Reply:
Added in 72adf435cb86a355b30437ff8559452fd9de3b0e.